packages/backend-common/src/reading/GoogleGcsUrlReader.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Storage } from '@google-cloud/storage';
+import {
+  ReaderFactory,
+  ReadTreeResponse,
+  SearchResponse,
+  UrlReader,
+} from './types';
+import getRawBody from 'raw-body';
+import {
+  GOOGLE_GCS_HOST,
+  readGoogleGcsIntegrationConfig,
+} from '@backstage/integration';
+
+const parseURL = (
+  url: string,
+): { host: string; bucket: string; key: string } => {
+  const { host, pathname } = new URL(url);
+
+  if (host !== GOOGLE_GCS_HOST) {
+    throw new Error(`not a valid GCS URL: ${url}`);
+  }
+
+  const [, bucket, ...key] = pathname.split('/');
+  return {
+    host: host,
+    bucket,
+    key: key.join('/'),
+  };
+};
+
+export class GoogleGcsUrlReader implements UrlReader {
+  static factory: ReaderFactory = ({ config, logger }) => {
+    if (!config.has('integrations.googleGcs')) {
+      return [];
+    }
+    const gcsConfig = readGoogleGcsIntegrationConfig(
+      config.getConfig('integrations.googleGcs'),
+    );
+    let storage: Storage;
+    if (!gcsConfig.clientEmail || !gcsConfig.privateKey) {
+      logger.warn(
+        'googleGcs credentials not found in config. Using default credentials provider.',
+      );
+      storage = new Storage();
+    } else {
+      storage = new Storage({
+        credentials: {
+          client_email: gcsConfig.clientEmail || undefined,
+          private_key: gcsConfig.privateKey || undefined,
+        },
+      });
+    }
+    const reader = new GoogleGcsUrlReader(storage);
+    const predicate = (url: URL) => url.host === GOOGLE_GCS_HOST;
+    return [{ reader, predicate }];
+  };
+
+  constructor(private readonly storage: Storage) {}
+
+  async read(url: string): Promise<Buffer> {
+    try {
+      const { bucket, key } = parseURL(url);
+
+      return await getRawBody(
+        this.storage.bucket(bucket).file(key).createReadStream(),
+      );
+    } catch (error) {
+      throw new Error(`unable to read gcs file from ${url}`);
```

Comment:
Probably best to forward the error as well here right? Or did you avoid it for a particular reason?

Reply:
The error could be something cryptic (something like "unable to read stream..." or "PEM error...") so when troubleshooting I usually prefer something that clearly explains what happened. WDYT?

Reply:
Oh, sorry you meant adding the error. Fixing...